### PR TITLE
Reorganize build scripts: define compile options declaratively

### PR DESCRIPTION
**Pull request by kmaehashi**
Part of #5722.

Currently compilation options for each module are adjusted in `preconfigure_modules` and `make_extensions` in adhoc way.
https://github.com/cupy/cupy/blob/master/install/cupy_builder/cupy_setup_build.py

This PR intends to allow defining options for each module in more declarative way.

### CODE REVIEW

**Comment by leofang** [comment]
Interesting! Would this help us reduce [overlinking](https://github.com/cupy/cupy/pull/6730#issuecomment-1136859919) in the future?

**Comment by kmaehashi** [comment]
Yes, this will also make it easy to adjust which module to link which shared libs.

File: install/cupy_builder/_features.py [hunk 1]
```diff
@@ -1,8 +1,81 @@
...
+        assert self._version != self._UNDETERMINED, 'not configured yet'
+        return self._version
+
+    def __contains__(self, key: Any) -> bool:
+        # TODO(kmaehashi): Remove this transient function.
+        if not isinstance(key, str):
+            return False
+        try:
+            self[key]
+        except AttributeError as e:
```
Comment (emcastillo) [inline install/cupy_builder/_features.py:47]:
```suggestion
        except AttributeError:
```

**Comment by emcastillo** [comment]
```


01:08:12.455576 STDOUT 1622] | ___________________ TestCheckVersion.test_check_cuda_version ___________________ |  
-- | -- | --
01:08:12.455578 STDOUT 1622] |   |  
01:08:12.455638 STDOUT 1622] | self = <install_tests.test_build.TestCheckVersion testMethod=test_check_cuda_version> |  
01:08:12.455640 STDOUT 1622] |   |  
01:08:12.455645 STDOUT 1622] | @pytest.mark.gpu |  
01:08:12.455663 STDOUT 1622] | @pytest.mark.skipif(test_hip, reason='For CUDA environment') |  
01:08:12.455665 STDOUT 1622] | def test_check_cuda_version(self): |  
01:08:12.455670 STDOUT 1622] | with self.assertRaises(RuntimeError): |  
01:08:12.455719 STDOUT 1622] | >           build.get_cuda_version() |  
01:08:12.455721 STDOUT 1622] | E           AttributeError: module 'cupy_builder.install_build' has no attribute 'get_cuda_version'

[01:08:12.455576 STDOUT 1622]](https://ci.preferred.jp/cupy.linux.cuda102/108429/#L5333)	___________________ TestCheckVersion.test_check_cuda_version ___________________	
[01:08:12.455578 STDOUT 1622]](https://ci.preferred.jp/cupy.linux.cuda102/108429/#L5334)		
[01:08:12.455638 STDOUT 1622]](https://ci.preferred.jp/cupy.linux.cuda102/108429/#L5335)	self = <install_tests.test_build.TestCheckVersion testMethod=test_check_cuda_version>	
[01:08:12.455640 STDOUT 1622]](https://ci.preferred.jp/cupy.linux.cuda102/108429/#L5336)		
[01:08:12.455645 STDOUT 1622]](https://ci.preferred.jp/cupy.linux.cuda102/108429/#L5337)	    @pytest.mark.gpu	
[01:08:12.455663 STDOUT 1622]](https://ci.preferred.jp/cupy.linux.cuda102/108429/#L5338)	    @pytest.mark.skipif(test_hip, reason='For CUDA environment')	
[01:08:12.455665 STDOUT 1622]](https://ci.preferred.jp/cupy.linux.cuda102/108429/#L5339)	    def test_check_cuda_version(self):	
[01:08:12.455670 STDOUT 1622]](https://ci.preferred.jp/cupy.linux.cuda102/108429/#L5340)	        with self.assertRaises(RuntimeError):	
[01:08:12.455719 STDOUT 1622]](https://ci.preferred.jp/cupy.linux.cuda102/108429/#L5341)	>           build.get_cuda_version()	
[01:08:12.455721 STDOUT 1622]](https://ci.preferred.jp/cupy.linux.cuda102/108429/#L5342)	E           AttributeError: module 'cupy_builder.install_build' has no attribute 'get_cuda_version'
```

**Comment by kmaehashi** [comment]
Test passed except for known ones.

**Review by emcastillo — APPROVED** [review]

**Pull request merged**